File: packages/core/src/image/gemini.ts
// Gemini 画像生成プロバイダー（Python `utils/image_provider/gemini.py` の移植）。
//
// `@google/genai` SDK でのリクエスト送信・参照画像の inlineData 化・base64 decode を
// 担う 1-attempt 契約の provider（#959）。リトライ・バックオフ・永続化は service 層
// （`service.ts` の `withRetry` + persist）が所有するため、ここでは SDK エラーを
// そのまま throw する（SAFETY / RECITATION は service 側の `shouldRetry` が
// non-retryable と判定する）。SDK client は注入され（テストは fake で差し替え）、
// ここはレスポンス decode と参照画像の inline 化に集中する。

import type { ImageGenerationRequest, ImageProvider } from "./base.ts";
import type { GeminiConfig } from "./config.ts";
import { isRecord } from "./internal.ts";
import { readReferenceFiles } from "./references.ts";

// `@google/genai` client の最小シェイプ（注入点の seam）。
interface GeminiClient {
  models: {
    generateContent(params: unknown): Promise<unknown>;
  };
}

/** GeminiImageProvider の注入依存。省略時は production default が使われる。 */
export interface GeminiProviderDeps {
  createClient: () => GeminiClient | Promise<GeminiClient>;
}

// `{ candidates: [{ content: { parts: [...] } }] }` から parts を取り出す。
// 想定外の形は空配列に落とし、「画像なしレスポンス」として throw 経路へ合流させる。
const extractParts = (response: unknown): unknown[] => {
  if (!isRecord(response) || !Array.isArray(response.candidates)) {
    return [];
  }
  const [first] = response.candidates;
  if (!isRecord(first) || !isRecord(first.content)) {
    return [];
  }
  return Array.isArray(first.content.parts) ? first.content.parts : [];
};

const inlineImageBytes = (part: unknown): Uint8Array | null => {
  if (
    isRecord(part) &&
    isRecord(part.inlineData) &&
    typeof part.inlineData.data === "string"
  ) {
    return new Uint8Array(Buffer.from(part.inlineData.data, "base64"));
  }
  return null;
};

const referenceMime = (path: string): string =>
  /\.jpe?g$/iu.test(path) ? "image/jpeg" : "image/png";

// 参照画像を base64 inlineData Part に変換し、末尾に prompt を付ける（参照なしは [prompt]）。
const buildContents = (
  prompt: string,
  references: readonly string[]
): unknown[] => {
  if (references.length === 0) {
    return [prompt];
  }
  const parts = readReferenceFiles(references).map(({ bytes, path }) => ({
    inlineData: {
      data: Buffer.from(bytes).toString("base64"),
      mimeType: referenceMime(path),
    },
  }));
  return [...parts, prompt];
};

const defaultCreateClient = async (): Promise<GeminiClient> => {
  const project = process.env.GOOGLE_CLOUD_PROJECT;
  if (!project) {
    throw new Error(
      "config: GOOGLE_CLOUD_PROJECT が未設定です。Vertex AI の project を指定してください"
    );
  }
  const { GoogleGenAI } = await import("@google/genai");
  // 画像系モデルは Vertex AI の global location のみサポート（genai_client.py 参照）。
  return new GoogleGenAI({
    location: "global",
    project,
    vertexai: true,
  }) as unknown as GeminiClient;
};

const defaultDeps = (): GeminiProviderDeps => ({
  createClient: defaultCreateClient,
});

/** Gemini API（Vertex AI 経由）で画像を 1 attempt 生成し、画像 bytes を返す。 */
export class GeminiImageProvider implements ImageProvider {
  readonly name = "gemini";
  // Gemini はアスペクト比を制限しない（branding/icon.png 用途で 1:1 等を許容）。
  readonly supportedAspectRatios: readonly string[] = [];

  private readonly config: GeminiConfig;
  private readonly deps: GeminiProviderDeps;

  // deps はテスト/サービス層が SDK client を差し替えるための注入点。省略時のみ
  // production default を使う（fallback ではなく明示的な DI 既定）。
  constructor(config: GeminiConfig, deps?: GeminiProviderDeps) {
    this.config = config;
    this.deps = deps ?? defaultDeps();
  }

  async generate(req: ImageGenerationRequest): Promise<Uint8Array> {
    const client = await this.deps.createClient();
    const references = req.references ?? [];
    // req.imageSize が空のときのみ config の既定解像度に委ねる（Python と同じ挙動）。
    const imageSize = req.imageSize || this.config.imageSize;
    const params = {
      config: {
        imageConfig: { aspectRatio: req.aspectRatio, imageSize },
        responseModalities: ["IMAGE", "TEXT"],
      },
      contents: buildContents(req.prompt, references),
      model: this.config.model,
    };

    // SDK エラー（SAFETY / RECITATION 含む）はそのまま伝播させる。
    const response = await client.models.generateContent(params);
    for (const part of extractParts(response)) {
      const bytes = inlineImageBytes(part);
      if (bytes) {
        return bytes;
      }
    }
    // 画像なしレスポンスは未 prefix Error（service 側で retryable と判定される）。
    throw new Error("gemini が画像なしレスポンスを返しました");
  }
}

File: src/youtube_automation/utils/image_provider/gemini.py
"""Gemini 画像生成プロバイダー。

`google.genai` SDK でのリクエスト送信・参照画像の bytes Part 化・SAFETY/RECITATION
即時失敗・指数バックオフリトライ・PNG/JPEG 保存を担う。
"""

from __future__ import annotations

import io
import time

from youtube_automation.domains.media.image import (
    RETRY_BACKOFF,
    RETRY_MAX,
    ImageGenerationRequest,
    ImageGenerationResult,
)
from youtube_automation.utils import cost_tracker
from youtube_automation.utils.genai_client import create_genai_client
from youtube_automation.utils.image_provider.composition import log_image_cost, persist_image
from youtube_automation.utils.image_provider.config import GeminiConfig


class GeminiImageProvider:
    """Gemini API（Vertex AI 経由）で画像を 1 枚生成して保存する。"""

    name = "gemini"
    # Gemini はアスペクト比を制限しない（branding/icon.png 用途で 1:1 等を許容）
    supported_aspect_ratios: tuple[str, ...] = ()

    def __init__(self, config: GeminiConfig) -> None:
        self._config = config

    def generate(self, req: ImageGenerationRequest) -> ImageGenerationResult:
        """req に従って画像を生成して保存する。成功時は ImageGenerationResult.success=True。"""
        from google.genai import types
        from PIL import Image as PILImage

        client = create_genai_client(location="global")
        model = self._config.model
        image_size = req.image_size or self._config.image_size
        aspect_ratio = req.aspect_ratio
        references = list(req.references)

        if references:
            contents = []
            for ref in references:
                ref_bytes = ref.read_bytes()
                mime = "image/jpeg" if ref.suffix.lower() in (".jpg", ".jpeg") else "image/png"
                contents.append(types.Part.from_bytes(data=ref_bytes, mime_type=mime))
            if self._config.variation_guard_enabled:
                variation_guard = (
                    "IMPORTANT: The reference image(s) above are for style and composition guidance ONLY. "
                    "Create an ORIGINAL image inspired by the reference — do NOT reproduce, copy, or closely "
                    "replicate the reference. Change the subject, colors, specific elements, and details "
                    "while keeping the general mood and layout style. The output must be clearly distinct "
                    "from the reference.\n\n"
                )
                contents.append(variation_guard + req.prompt)
            else:
                contents.append(req.prompt)
        else:
            contents = [req.prompt]

        save_as_png = req.output_path.suffix.lower() == ".png"

        for attempt in range(RETRY_MAX):
            try:
                ref_label = ""
                if references:
                    names = ", ".join(r.name for r in references)
                    ref_label = f" + 参照画像={names}"
                print(f"  [Submit] モデル={model} 解像度={image_size}{ref_label}")
                response = client.models.generate_content(
                    model=model,
                    contents=contents,
                    config=types.GenerateContentConfig(
                        response_modalities=["IMAGE", "TEXT"],
                        image_config=types.ImageConfig(
                            aspect_ratio=aspect_ratio,
                            image_size=image_size,
                        ),
                    ),
                )

                for part in response.parts:
                    if part.inline_data is not None:
                        saved_path = persist_image(
                            PILImage.open(io.BytesIO(part.inline_data.data)),
                            req.output_path,
                            save_as_png=save_as_png,
                        )
                        entry = log_image_cost(
                            model=model,
                            image_size=image_size,
                            aspect_ratio=aspect_ratio,
                            output_file=saved_path,
                            reference_count=len(references),
                        )
                        cost_tracker.print_last_report(entry)
                        return ImageGenerationResult(success=True, saved_path=saved_path)

                # 画像なしレスポンス
                text_parts = [p.text for p in response.parts if p.text]
                error_msg = " ".join(text_parts) if text_parts else "no image in response"
                print(f"  [Retry]  画像なし: {error_msg[:120]}")

            except Exception as e:
                error_msg = str(e)
                if "SAFETY" in error_msg.upper() or "RECITATION" in error_msg.upper():
                    print(f"  [Skip]   コンテンツポリシー違反: {error_msg[:120]}")
                    return ImageGenerationResult(success=False, saved_path=None)
                print(f"  [Retry]  attempt {attempt + 1}/{RETRY_MAX}: {error_msg[:120]}")

            if attempt < RETRY_MAX - 1:
                backoff = RETRY_BACKOFF[attempt]
                print(f"  [Wait]   {backoff}秒待機...")
                time.sleep(backoff)

        return ImageGenerationResult(success=False, saved_path=None)
